### PR TITLE
Bell/priority queue

### PR DESCRIPTION
**Pull request by songbell**
### Details:
 - fix the looping issue in auto->batching

### Tickets:
 - [*CVS-74542*](https://jira.devtools.intel.com/browse/CVS-74542)

### CODE REVIEW

File: src/inference/dev_api/threading/ie_thread_safe_containers.hpp [hunk 1]
```diff
@@ -48,6 +49,31 @@ template <typename T>
 using ThreadSafeQueue = tbb::concurrent_queue<T>;
 template <typename T>
 using ThreadSafeBoundedQueue = tbb::concurrent_bounded_queue<T>;
+template <typename T>
+class ThreadSafeBoundedPriorityQueue {
+public:
+    ThreadSafeBoundedPriorityQueue() = default;
+    bool try_push(T value) {
+        _mutex.lock();
```
Comment (tiger100256-hu) [inline src/inference/dev_api/threading/ie_thread_safe_containers.hpp:57]: maybe use unique lock，if(_capacity==0) always lock

Comment (songbell) [reply src/inference/dev_api/threading/ie_thread_safe_containers.hpp:57]: thanks for pointing out this bug, unique lock maybe too heavy for this piece of code, *push* itself is thread safe, fix it by adding unlock.

Comment (myshevts) [reply src/inference/dev_api/threading/ie_thread_safe_containers.hpp:57]: pls use the tbb::concurrent_bounded_queue

File: src/inference/include/ie/multi-device/multi_device_config.hpp [hunk 2]
```diff
@@ -42,5 +42,11 @@ DECLARE_MULTI_CONFIG_KEY(DEVICE_PRIORITIES);
  * is not 0, the network will be dispatched to the strongest device.
  */
 DECLARE_CONFIG_KEY(AUTO_NETWORK_PRIORITY);
+
+/**
+ * @brief temp config option, to disable auto batching explicitly
+ *
+ */
+DECLARE_CONFIG_KEY(ALLOW_AUTO_BATCHING);
```
Comment (myshevts) [inline src/inference/include/ie/multi-device/multi_device_config.hpp:50]: do we need this?

Comment (myshevts) [reply src/inference/include/ie/multi-device/multi_device_config.hpp:50]: (after the global config key with the same name\meaning is merged to the master)

**Comment by myshevts** [comment]
@songbell , this seems to be a very smart solution. 
it is worth to describe (maybe in the original ticket) whta is the root-cause.
From what I see in the code, this is becuase the requests may be scheduled in the arbitrary order, causing the (batched) requests in the auto-batching to starve.  So you fix that by priritizing by the underlying "batched request" number, so they group in the queue accordingly.

File: src/inference/dev_api/threading/ie_thread_safe_containers.hpp (hunk 1 above)
Comment (myshevts) [reply src/inference/dev_api/threading/ie_thread_safe_containers.hpp:57]: would making the _capacity atomic do the trick in a simpler way?

File: src/plugins/auto_batch/auto_batch.cpp [hunk 3]
```diff
@@ -349,6 +349,7 @@ InferenceEngine::IInferRequestInternal::Ptr AutoBatchExecutableNetwork::CreateIn
                         }
                         workerRequestPtr->_inferRequestBatched->StartAsync();
                     } else if ((status == std::cv_status::timeout) && sz) {
+                        std::cout << "timeout!!!" << std::endl;
```
Comment (myshevts) [inline src/plugins/auto_batch/auto_batch.cpp:352]: don't forget to remove when done :)

File: src/plugins/auto/executable_network.cpp [hunk 4]
```diff
@@ -296,18 +305,82 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+        _core->GetConfig(deviceNameWithoutBatch, CONFIG_KEY(PERFORMANCE_HINT)).as<std::string>() == CONFIG_VALUE(THROUGHPUT);
+    } catch (...) {
+        LOG_WARNING("[AUTOPLUGIN]not able to get performance hint config");
+    }
+    const auto& mode = config_with_batch.find(CONFIG_KEY(PERFORMANCE_HINT));
+    if ((!bThroughputEnabledInPlugin &&
+        (mode == config_with_batch.end() || mode->second != CONFIG_VALUE(THROUGHPUT))))
+        return;
+    try {
+            SetOptimalBatchNum(deviceNameWithoutBatch, config_with_batch);
```
Comment (myshevts) [inline src/plugins/auto/executable_network.cpp:361]: is this needed? (now when the autobatching is enabled transparently)?

File: src/plugins/auto/executable_network.cpp [hunk 5]
```diff
@@ -296,18 +305,82 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+                        }
+                        // round up to possible user's value
+                        _optimalBatchingRequestNum = (std::max)(_optimalBatchSize, requests);
+                        LOG_DEBUG("[AUTOPLUGIN]BATCHING %s:%ld", "optimal batch request num", _optimalBatchingRequestNum);
+                    } catch (...) {
+                        LOG_WARNING("[AUTOPLUGIN]get optimal infer request num for GPU auto-batch failed :%s");
+                    }
+            });
+}
+void MultiDeviceExecutableNetwork::TryApplyAutoBatching(AutoLoadContext& context) {
```
Comment (myshevts) [inline src/plugins/auto/executable_network.cpp:343]: is this func needed? (now when the autobatching is enabled transparently)?

File: src/inference/include/ie/multi-device/multi_device_config.hpp (hunk 2 above)
Comment (songbell) [reply src/inference/include/ie/multi-device/multi_device_config.hpp:50]: @myshevts This PR was originally test in together with auto batching before your #9724 is merge to master:) So a lot of code related to batching itself is not necessary. the more accurate PR targeting for this specific loop issue is https://github.com/openvinotoolkit/openvino/pull/9765
I'm testing if any other performance drop side effect with this now, though theoretically I think should not in our use case(with not that big data size in queue)

**Comment by wangleis** [comment]
> 

Will validate performance impact first

**Pull request closed without merging**